### PR TITLE
Replicate llama2 support

### PR DESCRIPTION
**Pull request by ruancomelli**
## Summary by Sourcery

Add support for using Replicate's Llama2 model as an alternative to OpenAI's GPT model by introducing a model factory that selects the appropriate LLM based on configuration

New Features:
- Add support for Replicate's Llama2 model as an alternative language model provider with appropriate system prompts and configuration

Enhancements:
- Refactor model initialization into a dedicated factory function to support multiple LLM providers
- Improve JSON parsing safety by introducing an intermediate variable for parsed results

Chores:
- Update emoji generation prompt to specifically request unicode emojis
- Remove unused imports from various files

### CODE REVIEW

**Comment by ruancomelli** [comment]
This is a benchmark review for experiment `reviews_for_yhkli`.
Run ID: `reviews_for_yhkli/benchmark_2024-10-29T11-27-46_v1-23-0-113-g9fada7a5a-dirty`.

This pull request was cloned from `https://github.com/ykhli/AI-tamago/pull/4`. _(Note: the URL is not a link to avoid triggering a notification on the original pull request.)_
<details>
<summary>Experiment configuration</summary>

```yaml
review_config:
  # User configuration for the review
  # - benchmark - use the user config from the benchmark reviews
  # - <value> - use the value directly
  user_review_config:
    enable_ai_review: true
    enable_rule_comments: false

    enable_complexity_comments: benchmark
    enable_security_comments: benchmark
    enable_tests_comments: benchmark
    enable_comment_suggestions: benchmark
    enable_functionality_review: benchmark

    enable_pull_request_summary: benchmark
    enable_review_guide: benchmark

    enable_approvals: false

  ai_review_config:
    # The model responses to use for the experiment
    # - benchmark - use the model responses from the benchmark reviews
    # - llm - call the language model to generate responses
    model_responses:
      comments_model: benchmark
      comment_validation_model: benchmark
      comment_suggestion_model: benchmark
      complexity_model: benchmark
      functionality_model: benchmark
      security_model: benchmark
      tests_model: benchmark
      pull_request_summary_model: benchmark
      review_guide_model: benchmark
      overall_comments_model: benchmark

# The pull request dataset to run the experiment on
pull_request_dataset:
- https://github.com/ykhli/AI-tamago/pull/25
- https://github.com/ykhli/AI-tamago/pull/22
- https://github.com/ykhli/AI-tamago/pull/21
- https://github.com/ykhli/AI-tamago/pull/19
- https://github.com/ykhli/AI-tamago/pull/18
- https://github.com/ykhli/AI-tamago/pull/17
- https://github.com/ykhli/AI-tamago/pull/16
- https://github.com/ykhli/AI-tamago/pull/15
- https://github.com/ykhli/AI-tamago/pull/14
- https://github.com/ykhli/AI-tamago/pull/13
- https://github.com/ykhli/AI-tamago/pull/12
- https://github.com/ykhli/AI-tamago/pull/11
- https://github.com/ykhli/AI-tamago/pull/10
- https://github.com/ykhli/AI-tamago/pull/9
- https://github.com/ykhli/AI-tamago/pull/8
- https://github.com/ykhli/AI-tamago/pull/7
- https://github.com/ykhli/AI-tamago/pull/6
- https://github.com/ykhli/AI-tamago/pull/5
- https://github.com/ykhli/AI-tamago/pull/4
- https://github.com/ykhli/AI-tamago/pull/3
- https://github.com/ykhli/AI-tamago/pull/2
- https://github.com/ykhli/AI-tamago/pull/1

# Questions to ask to label the review comments
review_comment_labels: []
# - label: correct
#   question: Is this comment correct?

# Benchmark reviews generated by running
#   python -m scripts.experiment benchmark <experiment_name>
benchmark_reviews:
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4792
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/414
- dataset_pull_request: https://github.com/ykhli/AI-tamago/pull/1
  review_pull_request: https://github.com/sourcery-ai-experiments/AI-tamago/pull/1

```

</details>

**Comment by SourceryAI** [comment]
## Reviewer's Guide by Sourcery

This PR adds support for using Replicate's Llama2 model as an alternative to OpenAI's GPT model. The implementation extracts model initialization into a separate utility function that checks environment variables to determine which model to use. It also includes some minor JSON formatting improvements and cleanup.

#### Sequence diagram for model initialization and interaction flow

```mermaid
sequenceDiagram
    participant C as Client
    participant A as API Route
    participant MF as Model Factory
    participant M as LLM Model

    C->>A: POST /api/interact
    A->>MF: getModel()
    alt LLM_MODEL = replicate_llama
        MF-->>A: Return Replicate Llama instance
    else default
        MF-->>A: Return OpenAI instance
    end
    A->>M: Process interaction
    M-->>A: Return response
    A-->>C: Send response
```

#### Class diagram showing the new model configuration structure

```mermaid
classDiagram
    class LlmModel {
        <<enumeration>>
        OpenAI
        ReplicateLlama
    }

    class ModelFactory {
        +getModel() BaseModel
    }

    class BaseModel {
        <<interface>>
    }

    class OpenAI {
        -modelName: string
        -openAIApiKey: string
        +verbose: boolean
    }

    class Replicate {
        -apiKey: string
        -model: string
        -input: object
    }

    BaseModel <|-- OpenAI
    BaseModel <|-- Replicate
    ModelFactory ..> BaseModel : creates
    note for ModelFactory "New utility class that
handles model selection
based on environment
variables"
```

### File-Level Changes

| Change | Details | Files |
| ------ | ------- | ----- |
| Added model selection functionality with support for Replicate's Llama2 | <ul><li>Created a new model utility module with model selection logic</li><li>Added environment variable-based model selection</li><li>Implemented Replicate Llama2 configuration with specific model version</li><li>Added system prompt configuration for JSON-only responses</li></ul> | `src/app/utils/model.ts`<br/>`src/app/api/interact/route.ts` |
| Improved JSON handling and formatting | <ul><li>Fixed JSON example format in prompt string</li><li>Refactored JSON parsing to use intermediate variable</li><li>Updated emoji prompt to specifically request unicode emojis</li></ul> | `src/app/api/interact/route.ts`<br/>`src/app/utils/interaction.ts` |
| Code cleanup and minor improvements | <ul><li>Removed unused imports</li><li>Fixed formatting and whitespace</li><li>Removed commented-out code</li></ul> | `src/app/page.tsx`<br/>`src/app/utils/config.ts` |

---

<details>
<summary>Tips and commands</summary>

#### Interacting with Sourcery

- **Trigger a new review:** Comment `@sourcery-ai review` on the pull request.
- **Continue discussions:** Reply directly to Sourcery's review comments.
- **Generate a GitHub issue from a review comment:** Ask Sourcery to create an
  issue from a review comment by replying to it.
- **Generate a pull request title:** Write `@sourcery-ai` anywhere in the pull
  request title to generate a title at any time.
- **Generate a pull request summary:** Write `@sourcery-ai summary` anywhere in
  the pull request body to generate a PR summary at any time. You can also use
  this command to specify where the summary should be inserted.

#### Customizing Your Experience

Access your [dashboard](https://app.sourcery.ai) to:
- Enable or disable review features such as the Sourcery-generated pull request
  summary, the reviewer's guide, and others.
- Change the review language.
- Add, remove or edit custom review instructions.
- Adjust other review settings.

#### Getting Help

- [Contact our support team](mailto:support@sourcery.ai) for questions or feedback.
- Visit our [documentation](https://docs.sourcery.ai) for detailed guides and information.
- Keep in touch with the Sourcery team by following us on [X/Twitter](https://x.com/SourceryAI), [LinkedIn](https://www.linkedin.com/company/sourcery-ai/) or [GitHub](https://github.com/sourcery-ai).

</details>

File: src/app/api/interact/route.ts [hunk 1]
```diff
@@ -59,8 +57,9 @@ export async function POST(req: Request) {
           .call({ stats: JSON.stringify(stats) })
           .catch(console.error);
         const { text } = result!;
-        const food = JSON.parse(text).food;
-        const rating = JSON.parse(text).rating;
+        const resultJson = JSON.parse(text);
```
Comment (SourceryAI) [inline src/app/api/interact/route.ts:60]: **suggestion:** Add error handling for JSON parsing operations

Consider wrapping the JSON.parse calls in a try/catch block to handle invalid JSON gracefully. Also, the non-null assertion on result could be replaced with proper null checking.

```suggestion
        let resultJson;
        try {
            resultJson = JSON.parse(text);
        } catch (error) {
            throw new Error('Invalid JSON format');
        }
```

File: src/app/utils/model.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
+
+import { OpenAI } from "langchain/llms/openai";
+import { Replicate } from "langchain/llms/replicate";
+
+enum LlmModel {
+    OpenAI = 'openai',
+    ReplicateLlama = 'replicate_llama',
+}
+export function getModel() {
```
Comment (SourceryAI) [inline src/app/utils/model.ts:9]: **issue (complexity):** Consider using a configuration map pattern to manage model initialization

The current implementation adds unnecessary complexity by mixing enum usage with string comparisons and scattered configuration. Here's a simpler approach using a model configuration map:

```typescript
const MODEL_CONFIGS = {
  replicate_llama: {
    create: () => new Replicate({
      apiKey: process.env.REPLICATE_API_TOKEN,
      model: "meta/llama13b-v2-chat:2d19859030ff705a87c746f7e96eea03aefb71f166725aee39692f1476566d48",
      input: {
        system_prompt: `You are a helpful assistant, that only communicates using JSON files and no other words.
          The expected output from you has to be: 
          {
              "text": {response}
          }`
      }
    }),
  },
  openai: {
    create: () => new OpenAI({
      modelName: "gpt-3.5-turbo-16k",
      openAIApiKey: process.env.OPENAI_API_KEY,
    }),
  }
};

export function getModel() {
  const modelType = (process.env.LLM_MODEL || "openai").toLowerCase();
  const config = MODEL_CONFIGS[modelType] || MODEL_CONFIGS.openai;

  console.debug(`Using ${modelType}`);
  return config.create();
}
```

This approach:
1. Eliminates the enum+if/else complexity in favor of a simple lookup
2. Groups model configuration with its initialization logic
3. Makes adding new models easier - just add a new entry to MODEL_CONFIGS
4. Maintains the same functionality while reducing code complexity

**Review by SourceryAI — COMMENTED** [review]
Hey @ruancomelli - I've reviewed your changes - here's some feedback:

**Overall Comments**:
- Consider improving error handling beyond console.error() - proper error handling and user feedback would make the application more robust
- The Replicate model ID should be moved to configuration rather than being hardcoded in the code

<details>
<summary>Here's what I looked at during the review</summary>

- 🟡 **General issues**: 1 issue found
- 🟢 **Security**: all looks good
- 🟢 **Testing**: all looks good
- 🟡 **Complexity**: 1 issue found
- 🟢 **Documentation**: all looks good
</details>

***

[LangSmith trace](https://smith.langchain.com/o/7d8518ad-6dd3-5c7b-96d2-757769296335/projects/p/88560225-5072-48bc-967c-dc54da53f265/r/124ce3fd-6521-4caa-a031-1a0e23f7b87e?poll=true)

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment and I'll use the feedback to improve your reviews.
</sub>

**Pull request closed without merging**